be/src/exprs/array_functions.cpp

```diff
@@ -1594,4 +1594,137 @@ StatusOr<ColumnPtr> ArrayFunctions::concat(FunctionContext* ctx, const Columns&
+    NullColumnPtr null_result = nullptr;
+
+    ArrayColumn* array_column = nullptr;
+    if (columns[0]->is_nullable()) {
+        is_nullable = true;
+        has_null = (columns[0]->has_null() || has_null);
+
+        const auto* src_nullable_column = down_cast<const NullableColumn*>(columns[0].get());
+        array_column = down_cast<ArrayColumn*>(src_nullable_column->data_column().get());
+        null_result = NullColumn::create(*src_nullable_column->null_column());
+    } else {
+        array_column = down_cast<ArrayColumn*>(src_column.get());
+    }
+
+    Int64Column* offset_column = nullptr;
+    if (columns[1]->is_nullable()) {
+        is_nullable = true;
+        has_null = (columns[1]->has_null() || has_null);
+
+        const auto* src_nullable_column = down_cast<const NullableColumn*>(columns[1].get());
+        offset_column = down_cast<Int64Column*>(src_nullable_column->data_column().get());
+        if (null_result) {
+            null_result = FunctionHelper::union_null_column(null_result, src_nullable_column->null_column());
+        } else {
+            null_result = NullColumn::create(*src_nullable_column->null_column());
+        }
+    } else {
+        offset_column =
+                down_cast<Int64Column*>(ColumnHelper::unpack_and_duplicate_const_column(chunk_size, columns[1]).get());
+    }
+
+    Int64Column* length_column = nullptr;
+    // length_column is provided.
+    if (columns.size() > 2) {
+        if (columns[2]->is_nullable()) {
+            is_nullable = true;
+            has_null = (columns[2]->has_null() || has_null);
+
+            const auto* src_nullable_column = down_cast<const NullableColumn*>(columns[2].get());
+            length_column = down_cast<Int64Column*>(src_nullable_column->data_column().get());
+            if (null_result) {
+                null_result = FunctionHelper::union_null_column(null_result, src_nullable_column->null_column());
+            } else {
+                null_result = NullColumn::create(*src_nullable_column->null_column());
+            }
+        } else {
+            length_column = down_cast<Int64Column*>(
+                    ColumnHelper::unpack_and_duplicate_const_column(chunk_size, columns[2]).get());
+        }
+    }
+
+    ArrayColumn* dest_data_column = nullptr;
+    if (columns[0]->is_nullable()) {
+        auto& dest_nullable_column = down_cast<NullableColumn&>(*dest_column);
+        dest_data_column = down_cast<ArrayColumn*>(dest_nullable_column.data_column().get());
+        auto& dest_null_data = dest_nullable_column.null_column_data();
+
+        dest_null_data = null_result->get_data();
+        dest_nullable_column.set_has_null(has_null);
+    } else {
+        dest_data_column = down_cast<ArrayColumn*>(dest_column.get());
+    }
+
+    if (columns.size() > 2) {
+        for (size_t i = 0; i < chunk_size; i++) {
+            _array_slice_item<true>(array_column, i, dest_data_column, offset_column->get(i).get_int64(),
```

Comment:
it seems that offset and length would often be constant value,  suggest to provide a specialization for handling this.

column->get(i) return Datum that wrapps an std::variant,  in fact, it is a union so the most wide type in this union decide the size of the union,  please make ensure that O3 optimization can eliminate transform int64->Datum->int64, if not, suggest to use Int64Column directly.

Reply:
Yeah, your are right, performance improve will in other PR, this PR just move code